enrichments/trace/internal/elastic/span.go

```diff
@@ -0,0 +1,294 @@
+		case semconv.AttributeURLFull:
+			s.isHTTP = true
+			// ignoring error as if parse fails then we don't want the url anyway
+			s.urlFull, _ = url.Parse(v.Str())
+		case semconv.AttributeURLScheme:
+			s.isHTTP = true
+			s.urlScheme = v.Str()
+		case semconv.AttributeURLDomain:
+			s.isHTTP = true
+			s.urlDomain = v.Str()
+		case semconv.AttributeURLPort:
+			s.isHTTP = true
+			s.urlPort = v.Int()
+		case semconv.AttributeURLPath:
+			s.isHTTP = true
+			s.urlPath = v.Str()
+		case semconv.AttributeURLQuery:
+			s.isHTTP = true
+			s.urlQuery = v.Str()
+		case semconv.AttributeRPCGRPCStatusCode:
+			s.isRPC = true
+			s.grpcStatus = codes.Code(v.Int()).String()
+		case semconv.AttributeRPCSystem:
+			s.isRPC = true
+			s.rpcSystem = v.Str()
+		case semconv.AttributeRPCService:
+			s.isRPC = true
+			s.rpcService = v.Str()
+		case semconv.AttributeDBStatement,
+			semconv.AttributeDBUser:
+			s.isDB = true
+		case semconv.AttributeDBName:
+			s.isDB = true
+			s.dbName = v.Str()
+		case semconv.AttributeDBSystem:
+			s.isDB = true
+			s.dbType = v.Str()
+		}
+		return true
+	})
+
+	// Ensure all dependent attributes are handled.
+	s.normalizeAttributes()
+
+	// Enrich the span depending on the nature of the span.
+	if isElasticTransaction(span) {
+		s.setTxnType(span)
+		s.setTxnResult(span)
+		s.setEventOutcome(span)
+	} else {
+		s.setEventOutcome(span)
+		s.setServiceTarget(span)
+	}
+}
+
+// normalizeAttributes sets any dependent attributes that
+// might not have been explicitly set as an attribute.
+func (s *Span) normalizeAttributes() {
+	if s.rpcSystem == "" && s.grpcStatus != "" {
+		s.rpcSystem = "grpc"
+	}
+}
+
+func (s *Span) setTxnType(span ptrace.Span) {
+	txnType := "unknown"
+	switch {
+	case s.isMessaging:
+		txnType = "messaging"
+		span.Attributes().PutStr("transaction.type", "messaging")
+	case s.isRPC, s.isHTTP:
+		txnType = "request"
+	}
+	span.Attributes().PutStr(AttributeTransactionType, txnType)
+}
+
+func (s *Span) setTxnResult(span ptrace.Span) {
+	var result string
+
+	if s.isHTTP && s.httpStatusCode > 0 {
+		switch i := s.httpStatusCode / 100; i {
+		case 1, 2, 3, 4, 5:
+			result = standardStatusCodeResults[i-1]
+		default:
+			result = fmt.Sprintf("HTTP %d", s.httpStatusCode)
+		}
+	}
+	if s.isRPC {
+		result = s.grpcStatus
+	}
+	if result == "" {
+		switch s.spanStatusCode {
+		case ptrace.StatusCodeError:
+			result = "Error"
+		default:
+			// default to success if all else fails
+			result = "Success"
+		}
+	}
+
+	span.Attributes().PutStr(AttributeTransactionResult, result)
+}
+
+func (s *Span) setEventOutcome(span ptrace.Span) {
+	// default to success outcome
+	outcome := "success"
+	switch {
+	case s.spanStatusCode == ptrace.StatusCodeError:
+		outcome = "failure"
+	case s.httpStatusCode >= http.StatusInternalServerError:
+		// TODO (lahsivjar): Handle GRPC status code? - not handled in apm-data
+		// TODO (lahsivjar): Move to HTTPResponseStatusCode? Backward compatibility?
+		outcome = "failure"
```

Comment:
Looks like there's a slight discrepancy with apm-data here: in apm-data we only the the outcome based on HTTP status code if the span status is unset.

Reply:
Fixed!